packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts

```diff
@@ -0,0 +1,270 @@
+import {
+  IOrderModuleService,
+  OrderChangeDTO,
+  OrderDTO,
+  PromotionDTO,
+} from "@medusajs/framework/types"
+import {
+  ApplicationMethodAllocation,
+  ChangeActionType,
+  MedusaError,
+  Modules,
+} from "@medusajs/framework/utils"
+import {
+  StepResponse,
+  WorkflowData,
+  WorkflowResponse,
+  createStep,
+  createWorkflow,
+  transform,
+  when,
+} from "@medusajs/framework/workflows-sdk"
+import { useRemoteQueryStep } from "../../common"
+import { deleteOrderChangeActionsStep } from "../steps/delete-order-change-actions"
+import { updateOrderChangesStep } from "../steps/update-order-changes"
+import { throwIfOrderChangeIsNotActive } from "../utils/order-validation"
+import { computeAdjustmentsForPreviewWorkflow } from "./order-edit/compute-adjustments-for-preview"
+
+/**
+ * The data to set the carry over promotions flag for an order change.
+ */
+export type SetCarryOverPromotionFlagWorkflowInput = {
+  /**
+   * The order change's ID.
+   */
+  order_change_id: string
+  /**
+   * Whether to carry over promotions to outbound exchange items.
+   */
+  carry_over_promotions: boolean
+}
+
+/**
+ * This step validates that the order change is an exchange and validates promotion allocation.
+ */
+export const validateCarryOverPromotionFlagStep = createStep(
+  "validate-carry-over-promotion-flag",
+  async function ({
+    orderChange,
+    order,
+    input,
+  }: {
+    orderChange: OrderChangeDTO
+    order: OrderDTO & { promotions?: PromotionDTO[] }
+    input: SetCarryOverPromotionFlagWorkflowInput
+  }) {
+    // Validate order change is active
+    throwIfOrderChangeIsNotActive({ orderChange })
+
+    // Validate order change is exchange
+    if (!orderChange.exchange_id) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Order change ${orderChange.id} is not an exchange.`
+      )
+    }
+
+    // we don't need to validate promotion since we will be resetting the adjustments
+    if (!input.carry_over_promotions) {
+      return
+    }
+
+    // Validate promotion allocation if promotions exist
+    if (order.promotions && order.promotions.length > 0) {
+      const invalidPromotions: string[] = []
+
+      for (const promotion of order.promotions) {
+        const applicationMethod = (promotion as any).application_method
+
+        if (!applicationMethod) {
+          continue
+        }
+
+        const allocation = applicationMethod.allocation
+        const type = applicationMethod.type
+
```

Comment:
self-review(todo): if allocation is other than `EACH` | `ACROSS` (e.g. `ONCE`) throw early